native-authentication/src/main/java/com/okta/nativeauthentication/NativeAuthenticationClient.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2022-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.nativeauthentication
+
+import com.okta.authfoundation.client.OidcClientResult
+import com.okta.authfoundation.client.internal.SdkVersionsRegistry
+import com.okta.authfoundation.credential.Token
+import com.okta.idx.kotlin.client.InteractionCodeFlow
+import com.okta.nativeauthentication.form.Form
+import com.okta.nativeauthentication.form.RetryFormBuilder
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+
+class NativeAuthenticationClient private constructor() {
+    companion object {
+        init {
+            SdkVersionsRegistry.register(SDK_VERSION)
+        }
+
+        fun create(
+            callback: Callback,
+            interactionCodeFlowFactory: suspend () -> OidcClientResult<InteractionCodeFlow>,
+        ): Flow<Form> {
+            suspend fun ProducerScope<Form>.start() {
+                when (val result = interactionCodeFlowFactory()) {
+                    is OidcClientResult.Error -> {
+                        RetryFormBuilder.emit(this) {
+                            start()
```

Comment:
Passing in producerScope to other functions might make it a bit harder to track emissions once more emitting functions are added. For future iterations, could we look into having only the channelFlow-owning class do the emissions, and the other classes emit events for the base class to consume? (to make it more of a unidirectional data flow)

Reply:
Yeah. I agree with this. I'll plan to keep it as is for now. But I'll keep this comment in mind as I'm refactoring when adding in future features.